File: filefn/python/filefn/processing/processors/compression.py
import gzip
import io
from typing import Any, Optional, List
from ..types import Processor, ProcessorInput, ProcessorResult, ProcessorOutputArtifact, CompressionConfig

SUPPORTED_MIME_TYPES = ['*/*']
UNSUPPORTED_ALGORITHM_ERROR = 'FILEFN_PROCESSING_UNSUPPORTED_COMPRESSION_ALGORITHM'
COMPRESSION_FAILED_ERROR = 'FILEFN_PROCESSING_COMPRESSION_FAILED'

class CompressionProcessor:
    def __init__(self, config: Optional[CompressionConfig] = None):
        config = config or CompressionConfig()
        self.algorithm = config.algorithm or 'gzip'
        self.level = config.level or 6
        self.name = 'compression'
        self.supportedMimeTypes = SUPPORTED_MIME_TYPES

    async def process(self, input: ProcessorInput, get_data: Any) -> ProcessorResult:
        if self.algorithm != 'gzip':
             return ProcessorResult(success=False, artifacts=[], error=UNSUPPORTED_ALGORITHM_ERROR)

        try:
            data = await get_data()
            
            out_buffer = io.BytesIO()
            with gzip.GzipFile(fileobj=out_buffer, mode='wb', compresslevel=self.level) as f:
                f.write(data)
            
            compressed_data = out_buffer.getvalue()
            
            artifacts = [ProcessorOutputArtifact(
                kind='compressed',
                data=compressed_data,
                mimeType='application/gzip',
                storageKey=f"{input.storageKey}.gz",
                metadata={
                    'algorithm': self.algorithm,
                    'level': self.level,
                    'originalSize': len(data),
                    'compressedSize': len(compressed_data),
                    'ratio': len(compressed_data) / len(data) if len(data) > 0 else 0
                }
            )]

            return ProcessorResult(success=True, artifacts=artifacts)

        except Exception:
            return ProcessorResult(success=False, artifacts=[], error=COMPRESSION_FAILED_ERROR)

def create_compression_processor(config: Optional[CompressionConfig] = None) -> Processor:
    return CompressionProcessor(config)

COMPRESSION_SUPPORTED_MIME_TYPES = SUPPORTED_MIME_TYPES

File: filefn/processing/src/processors/compression.ts
import { deflate, gzip } from 'node:zlib';
import { promisify } from 'node:util';
import type { Processor, ProcessorInput, ProcessorResult, CompressionConfig } from '../types.js';

const gzipAsync = promisify(gzip);
const deflateAsync = promisify(deflate);

const ALL_MIME_TYPES = ['*/*'];

export function createCompressionProcessor(config: CompressionConfig = {}): Processor {
  const { algorithm = 'gzip', level = 6 } = config;

  return {
    name: 'compression',
    supportedMimeTypes: ALL_MIME_TYPES,

    async process(input: ProcessorInput, getData: () => Promise<Uint8Array>): Promise<ProcessorResult> {
      try {
        const data = await getData();

        if (data.length === 0) {
          return {
            success: true,
            artifacts: [],
          };
        }

        const compressedData = await compress(data, algorithm, level);

        const compressionRatio = compressedData.length / data.length;

        if (compressionRatio >= 0.95) {
          return {
            success: true,
            artifacts: [],
          };
        }

        const extension = algorithm === 'gzip' ? 'gz' : 'deflate';
        const mimeType = algorithm === 'gzip' ? 'application/gzip' : 'application/octet-stream';

        return {
          success: true,
          artifacts: [
            {
              kind: `compressed-${algorithm}`,
              data: compressedData,
              mimeType,
              storageKey: `${input.storageKey}.${extension}`,
              metadata: {
                algorithm,
                level,
                originalSize: data.length,
                compressedSize: compressedData.length,
                compressionRatio,
                sourceFileId: input.fileId,
                sourceVersionId: input.versionId,
              },
            },
          ],
        };
      } catch (error) {
        return {
          success: false,
          artifacts: [],
          error: error instanceof Error ? error.message : 'Unknown compression error',
        };
      }
    },
  };
}

async function compress(data: Uint8Array, algorithm: 'gzip' | 'deflate', level: number): Promise<Uint8Array> {
  const options = { level };
  const buffer = Buffer.from(data);

  if (algorithm === 'gzip') {
    const result = await gzipAsync(buffer, options);
    return new Uint8Array(result);
  } else {
    const result = await deflateAsync(buffer, options);
    return new Uint8Array(result);
  }
}

export { ALL_MIME_TYPES as COMPRESSION_SUPPORTED_MIME_TYPES };
